File: 프로그래머스/2/389480. 완전범죄/완전범죄.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
const int INF = 987654321;

vector<int> dp;

int solution(vector<vector<int>> info, int n, int m) {
    int answer = INF;
    dp.resize(m, INF);
    dp[0] = 0;
    
    for(int i = 0; i < info.size(); i++) {
        vector<int> temp(m, INF);
        
        for(int b_evid = 0; b_evid < m; b_evid++) {
            if (dp[b_evid] == INF)
                continue;
            
            //A가 가져감
            if (dp[b_evid] + info[i][0] < n) {
                temp[b_evid] = min(temp[b_evid], dp[b_evid] + info[i][0]);
            }
            //B가 가져감
            if (b_evid + info[i][1] < m) {
                temp[b_evid + info[i][1]] = min(temp[b_evid + info[i][1]], dp[b_evid]);
            }
        }
        dp = temp;
    }
    
    for(int i = 0; i < m; i++) {
        answer = min(answer, dp[i]);
    }    
    
    return (answer == INF) ? -1 : answer;
}

File: 프로그래머스/2/389480. 완전범죄/완전범죄.java
import java.util.*;
class Solution {
    public int solution(int[][] info, int n, int m) {
        int INF = 100000;
        // info
        // 1 2 2
        // 2 3 1
        
        int len = info.length;
        int[][] dp = new int[len+1][m];
        for(int i=0; i<=len; i++) {
            Arrays.fill(dp[i], INF);
        }

        // dp[i][j] = A가 i번째 물건까지 훔쳤는데, B의 흔적이 j일 때의 A의 흔적
        dp[0][0] = 0;
        for(int i=1; i<=len; i++) {
            int a = info[i-1][0];
            int b = info[i-1][1];
            // i번째, A가 훔치거나 / B가 훔치거나
            for(int j=0; j<m; j++) {
                // 1) A가 훔칠 경우
                dp[i][j] = Math.min(dp[i-1][j] + a, dp[i][j]);
                // 2) B가 훔칠 경우
                if(j+b < m) 
                    dp[i][j+b] = Math.min(dp[i-1][j], dp[i][j+b]);
            }
        }
        
        int answer = INF;
        for(int i=0; i<m; i++)
            answer = Math.min(answer, dp[len][i]);
        
        return answer >= n ? -1 : answer;
    }
}